Add automatic media job type compatibility

incjo opened the pull request:
makes progress against #1065 by building on top of https://github.com/google/data-transfer-project/pull/1062 by further expanding support for automatically handling media jobs.

It allows not writing dedicated media adapters in cases where the existing photo/video functionality is satisfactory. It also provides automatic interoperability with destinations that only provide a Media adapter.

Main contributions:
- replaced `MediaToPhotoConversionImporter` and `MediaToPhotoConversionExporter` with generic versions that can support conversion between any two adapters: `AnyToAnyImporter` and `AnyToAnyExporter`
- added special case adapters `MediaExporterDecorator` and `MediaImporterDecorator` which combine existing photo and video adapters to seamlessly cover the media job type.
- added a compatibility layer `TransferCompatibilityProvider` (perhaps rename to `AdapterCompatibilityProvider`?) which  attempts to build a substitute adapter upon not finding a real one. E.g. when a media job can't find a media importer, it will combine existing photo and video importers to perform the job.

jzacsh commented on portability-spi-transfer/src/main/java/org/datatransferproject/spi/transfer/provider/converter/MediaToPhotoConversionImporter.java:
```diff
@@ -23,6 +23,7 @@
  */
 // TODO(#1065) fix primitives-obession causing us to key Providers on "PHOTOS" string rather
 // than underlying file types.
+@Deprecated
```
everywhere you add this annotation, can you comment also what the recommended alternative is, as an inline comment? eg:

```
@Deprecated // prefer FooBar#baz
```

jzacsh commented on portability-spi-transfer/src/test/java/org/datatransferproject/spi/transfer/provider/converter/MediaImporterDecoratorTest.java:
```diff
@@ -0,0 +1,86 @@
+
+    videos = List.of(
+        new VideoModel("v1", "", null, null, "d4", null, false),
+        new VideoModel("v2", "", null, null, "d5", "a3", false),
```
should the album params on these photo or video models match the params of the media albums above? Or are they intentionally mismatched for testing purpose?

incjo replied:
Ah, no reason. The particular data isn't relevant to the test.

jzacsh commented on portability-spi-transfer/src/test/java/org/datatransferproject/spi/transfer/provider/converter/MediaImporterDecoratorTest.java:
```diff
@@ -0,0 +1,86 @@
+    videosImporter = (id, ex, ad, data) -> ImportResult.OK;
+    mediaImporter = new MediaImporterDecorator<>(photosImporter, videosImporter);
+
+    albums = List.of(new MediaAlbum("album1", "name1", "desc1"),
```
(optional nitpick) suggestion to make these lines consistent with your others below

```suggestion
    albums = List.of(
        new MediaAlbum("album1", "name1", "desc1"),
```

jzacsh commented on portability-types-common/src/main/java/org/datatransferproject/types/common/models/media/MediaContainerResource.java:
```diff
@@ -61,11 +62,40 @@ public static PhotosContainerResource mediaToPhoto(MediaContainerResource mediaC
+   * Extracts a new VideosContainerResource from the relevant matching fields in a given
+   * MediaContainerResource.
+   */
+  public static VideosContainerResource mediaToVideo(MediaContainerResource mediaContainer) {
```
nit: consider adding test coverage of these two new static methods

jzacsh reviewed (changes requested):
sorry for the delay - I hadn't realized I hadn't submitted or finished my comments. Feel free to ping me in chat if my next replies aren't within a day.

jzacsh commented:
Do you have a pending PR somewhere that shows the final use of this? (eg: a PR that modifies a transfer extension?)

incjo commented on portability-spi-transfer/src/main/java/org/datatransferproject/spi/transfer/provider/converter/AnyToAnyExporter.java:
```diff
@@ -0,0 +1,54 @@
+   *                       sufficient.
+   */
+  public AnyToAnyExporter(Exporter<AD, From> exporter, Function<From, To> converter,
+      Function<ContainerResource, ContainerResource> inputConverter) {
```
Unfortunately, no. That's because an `Exporter<PhotosContainerResource>` takes `ExportInformation<ContainerResource>` as input to its `export` method. This container resource can be anything at all, e.g. `IdOnlyContainerResource`, `DateRangeContainerResource`, `PhotosContainerResource`, or even some private resource etc. For some reason, we're not enforcing strict typing in this case.

The `From` and `To` exporters might support different inputs and still be made compatible via the `AnyToAnyExporter`. For example, converting from `PhotosExporter` to `MediaExporter` might involve passing `Id` and `DateRange` resources unchanged and converting the `PhotosContainerResource` to `MediaContainerResource`.

incjo commented on portability-spi-transfer/src/main/java/org/datatransferproject/spi/transfer/provider/converter/AnyToAnyExporter.java:
```diff
@@ -0,0 +1,54 @@
+   *                       export call. It's required because exporters often support various
+   *                       container resources as part of export information. E.g. photo exporters
+   *                       support DateRangeContainerResource, while media adapters might only
+   *                       support MediaContainerResource. Oftentimes, Function.identity() is
```
That is fair, I just don't have any such usages at the moment. I actually had that for the `MicrosoftTransferExtension` while it didn't have a dedicated media adapter.

incjo commented:
> Do you have a pending PR somewhere that shows the final use of this? (eg: a PR that modifies a transfer extension?)

In our case, we're creating Media jobs without adding dedicated Media adapters. Essentially, we're leveraging the `TransferCompatibilityProvider` to automatically separate a `MediaContainerResource` into photos and videos, then use existing video and photo exporters and importers to do the remainder of the work while having the whole transfer happen under a single job.

This way we get to seamlessly support Media jobs without adding any new adapters, in our codebase or the destinations.

incjo commented on portability-spi-transfer/src/main/java/org/datatransferproject/spi/transfer/provider/TransferCompatibilityProvider.java:
```diff
@@ -0,0 +1,143 @@
+ */
+public class TransferCompatibilityProvider {
+
+  private static final String PHOTOS = "PHOTOS";
```
Cool, I'll pick it up. I've been frustrated this issue as well.

jzacsh commented:
hey @incjo did you accidentally revert this PR via a force-push? I'm not able to find the code we reviewed here and indeed if I click ["compare"](https://github.com/google/data-transfer-project/compare/6e97d8d56a0cb45f29386dd1c7bf0e46796311c5..ed1f42bc9a60b44ee85c2cd272758e108f6800cf) on this _"incjo force-pushed the `master` branch from `6e97d8d` to `ed1f42b`"_ it looks like your new code is gone (like adding `@Deprecated` lines, or your new classes); screenshot inlined below:
![Screenshot of the force-push notification on github and the 6e97d8d56a0cb45f29386dd1c7bf0e46796311c5..ed1f42bc9a60b44ee85c2cd272758e108f6800cf diff](https://user-images.githubusercontent.com/156228/187278988-3b0d4b20-161e-4f99-8e5b-f719d1c5fe9e.png)

incjo commented:
> hey @incjo did you accidentally revert this PR via a force-push? I'm not able to find the code we reviewed here and indeed if I click compare it looks like your new code is gone

Yep, for some reason Github decides to close the PR if you force push the wrong thing which is exactly what I did. Thanks Github. I don't have permissions to reopen the PR. I've addressed your feedback and will probably just make a new PR with the changes.
